pkg/controller/sts_handlers.go

```diff
@@ -116,10 +116,25 @@ func (c *Controller) AssumeRoleWithWebIdentityHandler(w http.ResponseWriter, r *
 		return
 	}
 
+	isSSTSAudience := false
+	for _, audience := range saAuthResult.Status.Audiences {
+		if audience == TokenReviewAudience {
+			isSSTSAudience = true
+		}
+	}
+
+	if !isSSTSAudience {
+		if !saAuthResult.Status.Authenticated {
+			writeSTSErrorResponse(w, true, ErrSTSAccessDenied, fmt.Errorf("Access denied: Invalid Token, audience '%s' not found", TokenReviewAudience))
+			return
+		}
+	}
+
```

Comment:
```suggestion
```

I don't think this test is needed, the credential would be denied by TokenReview if the audiences don't match. The only time you would need this is if you have multiple audiences and could get a partial match

Reply:
@bburky I think we do need to check this. The docs state ([source](https://dev-k8sref-io.web.app/docs/authentication/tokenreview-v1/#tokenreviewstatus)):
> A client of the TokenReview API that sets the `spec.audiences` field should validate that a compatible audience identifier is returned in the `status.audiences` field to ensure that the TokenReview server is audience aware.

However, I would add the following to also check the `Authenticated` status-field when the STS audience does match (there may be other reasons why validation didn't pass):
```suggestion
	if !isSSTSAudience {
		writeSTSErrorResponse(w, true, ErrSTSAccessDenied, fmt.Errorf("Access denied: Invalid token, audience '%s' not found", TokenReviewAudience))
		return
	}
	if !saAuthResult.Status.Authenticated {
		writeSTSErrorResponse(w, true, ErrSTSAccessDenied, "Access denied: Token not authenticted")
		return
	}
```

Reply:
"to ensure that the TokenReview server is audience aware" sounds like ancient versions of Kubernetes didn't support this (and would have ignored this field in the TokenReview `spec:`), so I guess you can use this to check the response and ensure it was validated. That's a guess though, I'd have to research more.

It certainly doesn't hurt to include this check, and does improves error logging too.

The code right below this diff already includes:
```go
	if !saAuthResult.Status.Authenticated {
		writeSTSErrorResponse(w, true, ErrSTSAccessDenied, fmt.Errorf("Access denied: Invalid Token"))
		return
	}
```

Reply:
It may be nice to include `saAuthResult.Status.Error` in these error messages though. The apiserver will already include details about why the token was denied.

Reply:
We want to know why the token was denied.

In the scenario that a user sends a valid JWT token but it does not include the audience `sts.min.io` the API would return that the token is invalid and mislead to belive Operator is having a bug, instead if we explicitly log "Access denied: Invalid Token, audience 'sts.min.io' not found" the issue is clear.

We don't want to disclouse too much info about the kubernetes cluster to applications the API error in `saAuthResult.Status.Error` might unveal a verbose error message vulnerability.